### PR TITLE
Git plugin

### PR DESCRIPTION
**Pull request by lukasz-golebiewski**

### CODE REVIEW

File: src/Frost/Effects/Git.hs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+module Frost.Effects.Git where
+
+import Data.String.Utils
+import Data.List
+import Polysemy
+import SimpleCmd.Git
+
+import qualified Data.Text as T
+
+data GitEffect m a where
```
Comment (rabbitonweb) [inline src/Frost/Effects/Git.hs]: please rename to `Git`

Comment (rabbitonweb) [inline src/Frost/Effects/Git.hs]: Please rename to `Git`

File: src/Frost/Effects/Git.hs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+runGitEffect :: Member (Lift IO) r => Sem (GitEffect ': r) a -> Sem r a
+runGitEffect = interpret $ \case
+  DevsList -> sendM getContributors
+
+getContributors :: IO [String]
+getContributors  = do
+    output <- git "log" ["--pretty=short", "-s"]
+    let s = T.split (=='\n') (T.pack output)
+    let r = nub $ filter (\x -> startswith  "Author:" (T.unpack x)) s
+    let f = map (\x -> drop (length "Author: ") (T.unpack x)) r
```
Comment (rabbitonweb) [inline src/Frost/Effects/Git.hs]: I think it is more common to use general `fmap` from `Functor` rather then `map` from list

File: src/Frost/Effects/Git.hs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+runGitEffect :: Member (Lift IO) r => Sem (GitEffect ': r) a -> Sem r a
+runGitEffect = interpret $ \case
+  DevsList -> sendM getContributors
+
+getContributors :: IO [String]
+getContributors  = do
+    output <- git "log" ["--pretty=short", "-s"]
+    let s = T.split (=='\n') (T.pack output)
+    let r = nub $ filter (\x -> startswith  "Author:" (T.unpack x)) s
```
Comment (rabbitonweb) [inline src/Frost/Effects/Git.hs]: when dealing with lambdas, it's almost always more readable to go point free I think
```
filter (startswith "Author:" . T.unpack) s
``` 
but it's just a matter of style I gues

File: src/Frost/GitPlugin.hs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+import Frost.Effects.Git
+import Frost.Plugin
+import Text.Pandoc
+import Polysemy
+import PolysemyContrib
+import Data.Map.Strict
+
+gitPlugin :: Member GitEffect r => Plugin r
+gitPlugin = Plugin "git:devs" (\_ -> return []) atm
```
Comment (rabbitonweb) [inline src/Frost/GitPlugin.hs]: You might consider `justMetaPlugin` smart constructor

**Comment by rabbitonweb** [comment]
@lukasz-golebiewski once you have that `Git` effect, writing a content plugin that will list contributors should be straight-forward :) And it's a plugin that you can easily write unit test for.

File: src/Frost/Effects/Git.hs (hunk 3 above)
Comment (rabbitonweb) [reply src/Frost/Effects/Git.hs]: I've used that converter for a long time to get a point free intuition http://pointfree.io/. Once you train your brain, you want to point free all the things

File: src/Frost/GitPlugin.hs (hunk 4 above)
Comment (lukasz-golebiewski) [reply src/Frost/GitPlugin.hs]: I think this should be a content plugin

Comment (rabbitonweb) [reply src/Frost/GitPlugin.hs]: or we could have both :)

File: src/Frost/GitPlugin.hs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+module Frost.GitPlugin where
```
Comment (rabbitonweb) [inline src/Frost/GitPlugin.hs]: Should be called more something like `ContributorsGitPlugin`

File: test/GitPluginSpec.hs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+spec :: Spec
+spec =
+  describe "GitPlugin" $ do
+    it "should substitute frost code blocks with content from the git plugin" $ do
+      -- given
+      let blocks = [ CodeBlock ("",["frost:git:devs"],[]) ""]
+      let pandoc = Pandoc nullMeta blocks
+      -- when
+      let res = run $ runGitPure ["Dev1", "Dev2"] $ substitute gitPlugin ""
```
Comment (rabbitonweb) [inline test/GitPluginSpec.hs]: instead, consider

```haskell
let res = substitute gitPlugin "" 
               & runGitPure ["Dev1", "Dev2"]
               & run
```

File: src/Frost/GitPlugin.hs (hunk 5 above)
Comment (lukasz-golebiewski) [reply src/Frost/GitPlugin.hs]: `GitContributorsPlugin`

File: test/GitPluginSpec.hs (hunk 6 above)
Comment (lukasz-golebiewski) [reply test/GitPluginSpec.hs]: Maybe next time ;-)

**Comment by rabbitonweb** [comment]
Resolves https://github.com/rabbitonweb/frost/issues/6

**Pull request merged**